include/ck_tile/ops/gemm/pipeline/gemm_universal_pipeline_ag_bg_cr_policy.hpp

```diff
@@ -80,45 +86,59 @@ struct UniversalGemmPipelineAgBgCrPolicy
             // kfold and mpair dimension is not always required.
             // more dimension in merge_transform increase the difficulty of generating immarg offset
             // for compiler.
-            constexpr auto M0 = get_warp_size() * Problem::BlockGemmShape::BlockWarps::at(I0);
-            constexpr auto M1 = MPerBlock / M0;
+            constexpr auto MThreads = get_warp_size() * Problem::BlockGemmShape::BlockWarps::at(I0);
+            static_assert(MThreads <= MPerBlock,
+                          "Make sure GEMM M tile (BlockTile[0]) is greater than or equal to "
+                          "(get_warp_size() * BlockWarps[0])");
 
-            constexpr auto KThreadWrite     = Problem::kBlockSize / M0;
-            constexpr auto K0PerThreadWrite = K0 / KThreadWrite;
-            constexpr auto KThreadRead      = 64 / WarpGemm::kM;
-            constexpr auto K0PerThreadRead  = K0 / KThreadRead;
+            constexpr auto MPerThread = MPerBlock / MThreads;
+            static_assert(MPerBlock == MThreads * MPerThread);
 
-            constexpr auto kfold =
-                (K1 * M0 * sizeof(ADataType) > 128) ? 1 : 128 / (K1 * M0 * sizeof(ADataType));
-            constexpr auto KThreadReadPerm =
-                (kfold * K0PerThreadWrite / K0PerThreadRead) > 1
-                    ? KThreadRead / (kfold * K0PerThreadWrite / K0PerThreadRead)
-                    : KThreadRead;
+            constexpr auto KPerThreadForWrite  = Problem::kBlockSize / MThreads;
+            constexpr auto K0PerThreadForWrite = KIterPerWarp / KPerThreadForWrite;
+            constexpr auto KPerThreadForRead   = 64 / WarpGemm::kM;
```

Comment:
Could you put the 64 as a static constexpr variable like "ThreadPerWave"? Also for the number 128. The rest of the code looks good to me.